File: _attachments/couch.js
function Couch(url) {
    this.url = url || Couch.guessDB();
}
Couch.prototype.urlFor = function (path, query) {
    if (path.join) {
        path = path.join("/");
    }
    if (query) {
        path += "?" + Couch._keys(query).map(function (key) {
            if (key[0] === '$') {
                return encodeURIComponent(key.slice(1)) + "=" + encodeURIComponent(JSON.stringify(query[key]));
            } else {
                return encodeURIComponent(key) + "=" + encodeURIComponent(query[key]);
            }
        }).join("&");
    }
    return this.url + "/" + path;
};
Couch.prototype.http = function (method, obj, path, query, callback) {
    var req = new XMLHttpRequest();
    req.open(method, this.urlFor(path, query), Boolean(callback));
    req.setRequestHeader("Content-Type", "application/json");
    req.send(JSON.stringify(obj));
    if (callback) {
        req.onreadystatechange = function() {
            if (req.readyState === (req.DONE || 4)) {
                callback(req.status, JSON.parse(req.responseText));
            }
        }
    } else {
        req.response = JSON.parse(req.responseText);
    }
    return req;
};
Couch.prototype.get = function (path, query, callback) {
    if (callback) {
        this.http("GET", null, path, query, function (status, result) {
            callback((status === 200) ? result : null);
        });
    } else {
        var req = this.http("GET", null, path, query);
        return (req.status === 200) ? req.response : null;
    }
};

Couch.prototype.read = function (id) {
    var req = this.http("GET", null, id);
    if (req.status === 404) {
        return null;
    } else if (req.status !== 200) {
        throw Error(req.statusText);
    }
    return req.response;
}
Couch.prototype.write = function (doc) {
    var req = this.http("PUT", doc, doc._id);
    if (req.status !== 201) {
        throw Error(req.statusText);
    }
    doc._rev = req.response.rev;
}
Couch.prototype.remove = function (doc) {
    var req = this.http("DELETE", null, doc._id, {'rev':doc._rev});
    if (req.status !== 200) {
        throw Error(req.statusText);
    }
    doc._deleted = true;
    doc._rev = req.response.rev;
}


// helper needed under older JavaScript engines
Couch._keys = Object.keys || function (obj) {
    var keys = [], key;
    for (key in obj) if (obj.hasOwnProperty(key)) {
        keys.push(key);
    }
    return keys;
};

Couch.guessDB = function () {
    var x;
    if (location.protocol === "file:") {
        return "http://localhost:5984/dev";
    } else if ((x = location.href.indexOf("/_design")) !== -1) {
        return location.href.slice(0, x);
    } else {
        return "../..";
    }
}

Couch.REF_TYPE = "net.stemstorage.reference";

Couch.makeRef = function (doc, denormalize) {
    var reference = {};
    reference[Couch.REF_TYPE] = true;
    reference._id = doc._id;
    if (denormalize) denormalize.forEach(function (field) {
        reference[field] = doc[field];
    }), (reference._rev = doc._rev);
    return reference;
};

File: _attachments/scripts/couch.py
'''Simple classes for talking with CouchDB, as a client or as an external process'''

import sys
import json
import httplib
import urllib, urlparse

REF_TYPE = "testtype-reference"


class Database(object):
    def __init__(self, url):
        self.url = url
    
    def url_for(self, path, query=None):
        if type(path) not in (str, unicode):
            path = '/'.join(path)
        
        if query:
            path += "?" + urllib.urlencode([
                (key[1:], json.dumps(value)) if key[0] == '$' else (key, value) for (key,value) in query.iteritems()
            ])
        
        return "%s/%s" % (self.url, path)
    
    def http(self, method, obj, path, query=None):
        url = self.url_for(path, query)
        url_parts = urlparse.urlsplit(url)
        
        Con = httplib.HTTPSConnection if url_parts.scheme == 'https' else httplib.HTTPConnection
        conn = Con(url_parts.netloc)
        conn.request(method.upper(), url, json.dumps(obj) if obj else None, {'Content-Type':"application/json"})
        resp = conn.getresponse()
        return resp.status, json.loads(resp.read())
    
    def get(self, path='', query=None):
        status, result = self.http('GET', None, path, query)
        return result if status == 200 else None
    
    def read(self, doc_id):
        status, result = self.http('GET', None, doc_id)
        if status != 200:
            raise IOError(status, result['error'], doc_id)
        return result
    
    def write(self, doc):
        status, result = self.http('PUT', doc, doc['_id'])
        if status != 201:
            raise IOError(status, result['error'], doc['_id'])
        doc['_rev'] = result['rev']
    
    def remove(self, doc):
        status, result = self.http('DELETE', None, doc['_id'], {'rev':doc['_rev']})
        if status != 200:
            raise IOError(status, result['error'], doc['_id'])
        doc['_deleted'] = True
        doc['_rev'] = result['rev']

def make_ref(doc, denormalize=None):
    reference = {}
    reference[REF_TYPE] = True
    reference['_id'] = doc['_id']
    if denormalize is not None:
        for field in denormalize:
            reference[field] = doc[field]
        reference['_rev'] = doc['_rev']
    return reference



# see http://wiki.apache.org/couchdb/ExternalProcesses for configuration instructions
# and http://www.davispj.com/2010/09/26/new-couchdb-externals-api.html for The Future (?)
class External(object):
    def run(self):
        line = sys.stdin.readline()
        while line:
            try:
                response = self.process(json.loads(line))
            except Exception as e:
                response = {'code':500, 'json':{'error':True, 'reason':"Internal error processing request (%s)" % e}}
            sys.stdout.write("%s\n" % json.dumps(response))
            sys.stdout.flush()
            line = sys.stdin.readline()
    
    def process(self, req):
        return {'json':{'ok':True}}
